File: src/main/java/com/circleescape/server/model/PolarCoordinates.java
package com.circleescape.server.model;

import lombok.Value;

import static com.circleescape.server.model.MathUtils.*;

import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

import org.apache.commons.math3.geometry.euclidean.twod.Vector2D;

/***
 * immutable
 * @param r
 *
 */
@Value
public class PolarCoordinates {

	@NotNull
	@Min(0)
	private double r;
	
	@NotNull
	private double theta;
	
	/**
	 * Negative r throws illegal arguments exception
	 * Theta will be normalized s.t. 0 <= theta < TAU
	 * @param r
	 * @param theta
	 */
	public PolarCoordinates(double r, double theta) {
		if (r<0)
			throw new IllegalArgumentException("Radius can't be negative.");
		
		this.r = r;
		
		if ( theta < 0 ) {
			theta = TAU + theta % TAU;
		} else if ( TAU < theta ) {
			theta = theta % TAU;
		}
		this.theta = theta;		
	}
	
	public PolarCoordinates normalize(double r) {
		
		double normalizedR = Math.min(this.r, r);
		
		return new PolarCoordinates(normalizedR, theta);
	}
	
	public double distance(PolarCoordinates pc) {
		return Math.sqrt(Math.pow(r, 2) + Math.pow(pc.r, 2) - 2 * r * pc.r * Math.cos(theta - pc.theta));
	};
	
	public PolarCoordinates add(PolarCoordinates polarDiff) {
		return fromCartesian(toCartesian().add(polarDiff.toCartesian()));
	}
	
	public Vector2D toCartesian() {
		return new Vector2D(Math.cos( theta ) * r, 
		                    Math.sin( theta ) * r); 
	}
	
	private PolarCoordinates fromCartesian(Vector2D v) {
		double r_ = Math.sqrt( v.getX() * v.getX() + v.getY() * v.getY() );
		double theta_ = Math.atan2(v.getY(), v.getX());
		
		return new PolarCoordinates(r_, theta_);
	}
}

File: src/main/kotlin/com/circleescape/server/model/PolarCoordinates.kt
package com.circleescape.server.model

import org.apache.commons.math3.geometry.euclidean.twod.Vector2D
import javax.validation.constraints.Min
import kotlin.math.*

/***
 * immutable
 * @param r
 */
class PolarCoordinates(r: Double, theta: Double) {
    val r: @Min(0) Double
    val theta: Double

    /**
     * Negative r throws illegal arguments exception
     * Theta will be normalized s.t. 0 <= theta < TAU
     * @param r
     * @param theta
     */
    init {
        require(r >= 0) { "Radius can't be negative." }
        this.r = r
        var theta = theta % MathUtils.TAU
        if (theta < 0) {
            theta += MathUtils.TAU
        }
        this.theta = theta
    }

    fun normalize(r: Double): PolarCoordinates {
        val normalizedR = Math.min(this.r, r)
        return PolarCoordinates(normalizedR, theta)
    }

    fun distance(pc: PolarCoordinates): Double {
        return sqrt(r.pow(2.0) + pc.r.pow(2.0) - 2 * r * pc.r * cos(theta - pc.theta))
    }

    fun add(polarDiff: PolarCoordinates): PolarCoordinates {
        return fromCartesian(toCartesian().add(polarDiff.toCartesian()))
    }

    fun toCartesian(): Vector2D {
        return Vector2D(
            cos(theta) * r,
            sin(theta) * r
        )
    }

    private fun fromCartesian(v: Vector2D): PolarCoordinates {
        val r_ = sqrt(v.x * v.x + v.y * v.y)
        val theta_ = atan2(v.y, v.x)
        return PolarCoordinates(r_, theta_)
    }
}